File: week-02/day-2/src/Matchmaking.java
import java.util.*;

public class Matchmaking {

  public static void main(String... args) {
    ArrayList<String> girls = new ArrayList<String>(
        Arrays.asList("Eve", "Ashley", "Claire", "Kat", "Jane"));
    ArrayList<String> boys = new ArrayList<String>(
        Arrays.asList("Joe", "Fred", "Tom", "Todd", "Neef", "Jeff"));

    // Write a method that joins the two lists by matching one girl with one boy into a new list
    // If someone has no pair, he/she should be the element of the list too
    // Exepected output: "Eve", "Joe", "Ashley", "Fred"...

    System.out.println(makingMatches(girls, boys));
  }

  private static ArrayList<String> makingMatches(
      ArrayList<String> arrList1,
      ArrayList<String> arrList2
  ) {
    /**
     * This function takes 2 ArrayLists of String as parameters and
     *  merges their elements 1 by 1, starting with the first parameter.
     *
     * If the ArrayLists are not the same size, the function just puts
     *  the remaining element at the end of the result.
     *
     *
     * @param arrList1  first parameter - ArrayList<String>, EVEN indices
     * @param arrList2  second parameter - ArrayList<String>, ODD indices
     *
     * @return the merged ArrayList<String> of the parameters
     */

    ArrayList<String> matchList = new ArrayList<>();

    // set iterators
    Iterator<String> iterArrList1 = arrList1.iterator();
    Iterator<String> iterArrList2 = arrList2.iterator();

    while (iterArrList1.hasNext() || iterArrList2.hasNext()) {
      if (iterArrList1.hasNext()) {
        matchList.add(iterArrList1.next());
      }
      if (iterArrList2.hasNext()) {
        matchList.add(iterArrList2.next());
      }
    }
    return matchList;
  }
}

File: python_solutions/week-02/day-2/data_structures/matchmaking.py
import itertools

# Write function that joins the two lists by matching one girl with one boy
# in a new list
# If someone has no pair, he/she should be the element of the list too
# Expected output: ["Eve", "Joe", "Ashley", "Fred"...]

girls = ["Eve", "Ashley", "Claire", "Kat", "Jane"]
boys = ["Joe", "Fred", "Tom", "Todd", "Neef", "Jeff"]


def making_matches(girls: list, boys: list):
    result = list()
    
    for girl_and_boy in itertools.zip_longest(girls, boys):
        if girl_and_boy[0]:
            result.append(girl_and_boy[0])
        if girl_and_boy[1]:
            result.append(girl_and_boy[1])
    
    return result


if __name__ == '__main__':
    print(making_matches(girls, boys))
